hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java

```diff
@@ -456,167 +688,3155 @@ public void bindParameterValue(
+		appendSql( tableGroupJoin.getJoinedGroup().getPrimaryTableReference().getIdentificationVariable() );
+		appendSql( '.' );
+		//TODO: pretty sure the typecast to Loadable is quite wrong here
+		appendSql( ( (Loadable) tableGroupJoin.getJoinedGroup().getModelPart() ).getIdentifierColumnNames()[0] );
+	}
+
+	@Override
+	public void visitTableReference(TableReference tableReference) {
+		// nothing to do... handled via TableGroup#render
+	}
+
+	@Override
+	public void visitTableReferenceJoin(TableReferenceJoin tableReferenceJoin) {
+		// nothing to do... handled within TableGroup#render
+	}
+
+
+	// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+	// Expressions
+
+	@Override
+	public void visitColumnReference(ColumnReference columnReference) {
+		if ( dmlTargetTableAlias != null && dmlTargetTableAlias.equals( columnReference.getQualifier() ) ) {
+			// todo (6.0) : use the Dialect to determine how to handle column references
+			//		- specifically should they use the table-alias, the table-expression
+			//			or neither for its qualifier
+
+			// for now, use the unqualified form
+			appendSql( columnReference.getColumnExpression() );
+		}
+		else {
+			appendSql( columnReference.getExpressionText() );
+		}
+	}
+
+	@Override
+	public void visitExtractUnit(ExtractUnit extractUnit) {
+		appendSql( getDialect().translateExtractField( extractUnit.getUnit() ) );
+	}
+
+	@Override
+	public void visitDurationUnit(DurationUnit unit) {
+		appendSql( getDialect().translateDurationField( unit.getUnit() ) );
+	}
+
+	@Override
+	public void visitFormat(Format format) {
+		final String dialectFormat = getDialect().translateDatetimeFormat( format.getFormat() );
+		appendSql( "'" );
+		appendSql( dialectFormat );
+		appendSql( "'" );
+	}
+
+	@Override
+	public void visitStar(Star star) {
+		appendSql( "*" );
+	}
+
+	@Override
+	public void visitTrimSpecification(TrimSpecification trimSpecification) {
+		appendSql( " " );
+		appendSql( trimSpecification.getSpecification().toSqlText() );
+		appendSql( " " );
+	}
+
+	@Override
+	public void visitCastTarget(CastTarget castTarget) {
+		appendSql(
+				getDialect().getCastTypeName(
+						castTarget.getExpressionType(),
+						castTarget.getLength(),
+						castTarget.getPrecision(),
+						castTarget.getScale()
+				)
+		);
+	}
+
+	@Override
+	public void visitDistinct(Distinct distinct) {
+		appendSql( "distinct " );
+		distinct.getExpression().accept( this );
+	}
+
+	@Override
+	public void visitParameter(JdbcParameter jdbcParameter) {
+		if ( inlineParameters ) {
+			renderExpressionAsLiteral( jdbcParameter, jdbcParameterBindings );
+		}
+		else {
+			appendSql( PARAM_MARKER );
+
+			parameterBinders.add( jdbcParameter.getParameterBinder() );
+			jdbcParameters.addParameter( jdbcParameter );
+		}
+	}
+
+	@Override
+	public void render(SqlAstNode sqlAstNode, SqlAstNodeRenderingMode renderingMode) {
+		switch ( renderingMode ) {
+			case NO_PLAIN_PARAMETER:
```

Comment:
Out of curiosity, why not just cast?  

If you you really think the concat / +0 approach is better, then maybe just as a better fallback to calling `#renderExpressionAsLiteral`.  

```
if ( sqlAstNode instanceof JdbcParameter ) {
    final JdbcMapping jdbcMapping = ...;

    if ( jdbcMapping.getSqlTypeDescriptor().isNumber() ) {
        ...
    }
    else if ( jdbcMapping.getSqlTypeDescriptor().isString() ) {
        ...
    }
    else {
        try {
            final String targetType = dialect.getRawTypeName( 
                    jdbcMapping.getSqlTypeDescriptor().getJdbcTypeCode() 
            );
            appendSql( "cast(? as " );
            appendSql( targetType );
            appendSql( ")" );
        }
        catch( Exception e ) {
            // the Dialect did not know a proper database type for the code...
            renderExpressionAsLiteral();
        }
}
```

If the intention really is to avoid inlining the parameters as much as possible, this helps with that

Reply:
You are right, I adapted that. Thanks!